keps/sig-scheduling/3370-native-podgroup-api/README.md

```diff
@@ -0,0 +1,1105 @@
+	// successfully, but the scheduling fails because minMember cannot be met.
+	ScheduledAvailable int32
+
+	// ScheduledUnavailable is the number of pods in the subset that can't be 
+	// scheduled successfully.
+	ScheduledUnavailable int32
+
+	// Running is the number of pods targeted by this subset
+	// with a Running phase.
+	Running int32
+}
+```
+
+ After the pod group is scheduled for the first time, the condition `PodGroupScheduled` is
+ added by the `pod-group-controller` in kube-controller-manager. Then `pod-group-controller`
+ will reconcile the status of pod group. The workload operator can watch the pod group status,
+ for example, when a pod is deleted and the total running count is less than `MinMember`. The
+ workload operator can determine if it needs to evict the whole group or just create a new pod.
+
+
+### User workflow
+1. The user creates a PodGroup object.
+2. The user creates a workload (Deployment, Job, etc.) and associates .spec.podGroup to the previously created PodGroup name.
+3. The user watched the status of PodGroup. Pods associated with this PodGroup are expected to be co-scheduled if resources are adequate.
+4. The user waits for the PodGroup to be complete and then deletes it.
+
+### Scheduler
+#### QueueSort
+In order to maximize the chance that the pods which belong to the same `PodGroup` to be scheduled consecutively, we need to implement a customized `QueueSort` plugin to sort the Pods properly.
+
+Firstly, we will inherit the default in-tree PrioritySort plugin so as to honor .spec.priority to ensure high-priority Pods are always sorted ahead of low-priority ones.
+
+Secondly, if two Pods hold the same priority, the sorting precedence is described as below:
+
+- If they are both regularPods (without pod.spec.podGroup), compare their 
+`InitialAttemptTimestamp` field: the Pod with earlier `InitialAttemptTimestamp` is positioned 
+ahead of the other.
+  
+- If one is regularPod and the other is pgPod (with pod.spec.podGroup), compare regularPod's
+`InitialAttemptTimestamp` with the `creationtime` of pgPod's PodGroup: the Pod with earlier 
+timestamp is positioned ahead of the other.
+  
+- If they are both pgPods:
+  - Compare their pod group's `CreationTimestamp`: the Pod in pod group which has earlier timestamp is positioned ahead of the other.
+  - If their `CreationTimestamp` is identical, order by their UID of PodGroup: a Pod with lexicographically greater UID is scheduled ahead of the other Pod. (The purpose is to tease different PodGroups with the same `CreationTimestamp` apart, while also keeping Pods belonging to the same PodGroup back-to-back)
+
+#### PreFilter
+In PreFilter phase, we need to check if pod.spec.podGroup is nil firstly. If so, return 
+success. If not, check if the pod group exists in the scheduler's internal cache and pod.spec.
+podGroup.subset is matched in pod group, if not, reject the pod in PreFilter.
```

Comment:
In PreFilter, it does some preliminary checks in the following sequences:

- If the pod doesn't carry `.spec.podGroup`, returns `Success` immediately. Optionally, plumb a key-value pair in CycleState so as to be leveraged in `Permit` phase later.
- If the pod carries `.spec.podGroup`, verify if the PodGroup exists or not:
  - If not, return `UnschedulableAndUnresolvable`.
  - If yes, verify if the quorum (MinMember) has been met. If met, return `Success`; otherwise return `UnschedulableAndUnresolvable`.

Reply:
Done